metal: TRI, FILL, EXPM1, SOFTPLUS

gabe-l-hart opened the pull request:
## Description

EDIT: This PR has now been updated to match the kernel signatures in `master` and https://github.com/ggml-org/llama.cpp/pull/17584. CUMSUM was already added in https://github.com/ggml-org/llama.cpp/pull/17305 and there are several more ops added in #17063, so this PR now adds `TRI`, `FILL`, `EXPM1`, and `SOFTPLUS` for `Metal`.

---

This PR builds on some of the work by @pwilkin in #16095 and extends the CPU implementations of `CUMSUM` and `TRI` to `Metal` and `CUDA`. It also extends type support to `F16` and `BF16`.

The goal of this PR is to establish these two ops in the interest of both the `DELTA_NET` op for Qwen3-Next and the chunked implementation of the State Space Duality form of `SSM_SCAN` for faster prefill.

I'm putting this up for review now in case it helps with the Qwen3-Next work and to get feedback on the kernels. I'm quite novice at kernel development, so I suspect others may find significant optimizations for both Metal and CUDA.

gabe-l-hart commented:
Yikes, looks like some alternate platforms that will need to be handled. I'll dig through some of these failures

pwilkin commented:
> My opinion is that we should assert that the CPU implementation is correct before we move towards reviewing and merging this PR.

I think @gabe-l-hart moved the TRI and CUMSUM CPU implementation to this PR as well, so I guess it's a question of adding some testcases?

Those ops aren't very hard and as far as basic correctness goes I think I've verified them quite extensively during my fights with Qwen3 Next.

I've mimicked the basic logic for CUMSUM to be the same as SUM_ROWS, so it's basically always done on the first dimension.

gabe-l-hart commented:
@giuseppe I saw that yesterday! That support will help a ton. The SSD formulation is an additional optimization on top of the recurrent formulation that should have big benefits for prefill with long context. It's mathematically equivalent, but much more efficient. The challenge I'm working through right now is how best to decompose the problem to minimize impact. One option is to write it as a sub-graph composed of smaller ops that is used when the sequence length is > 1, but this would have two problems: 

1. It would break the ability to reuse the graph objects across generation steps

2. It would make backends that don't support the primitive ops (CUMSUM and TRI) fall back to CPU (exactly what you're trying to fix in your PR for SSM_SCAN)

The alternative is to implement it inside a single backend's SSM_SCAN. This has the advantage of being self-contained so it can be done on a per-backend basis, but it has the inverse problem of _requiring_ each backend to implement it separately in order to get the performance boost. It's also much harder to write as a single kernel since it involves allocating temporary tensors of different size than the input or output tensors.

fat-tire commented:
Dumb question-- but can the CUDA gated deltanet code be optimized via [Vidrial](https://manifestai.com/articles/what-is-vidrial/index.html)? It supposedly does autotuning by looking at various configurations to find an ideal configuration... (?)

gabe-l-hart commented:
@pwilkin if you're on a roll, you're welcome to go for it. I can get to it next week sometime if you don't get to it first.

pwilkin commented:
@gabe-l-hart Aight, might just do the CUDA kernels since I have neither the know-how nor the ability to test Metal kernels ;)

gabe-l-hart commented:
Deal, I can definitely tackle those

wsbagnsv1 commented:
> @gabe-l-hart Aight, might just do the CUDA kernels since I have neither the know-how nor the ability to test Metal kernels ;)

Im open to use my framework to optimize the cuda ones once again after your initial implementation, though I dont have metal hardware to test the metal ones either 😅

gabe-l-hart committed:
feat(wip): Port initial TRI impl from pervious work

The kernel does not work and is not optimized, but the
code compiles and runs, so this will be the starting point
now that the core op has been merged.

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Remove argument for constant val override

This was added in the original draft, but later removed. With this, the
kernel now passes tests.

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Move the ttype conditional to templating to avoid conditional in kernel

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov commented on ggml/src/ggml-metal/ggml-metal-impl.h:
```diff
@@ -831,6 +831,25 @@ typedef struct {
+    int64_t  ne00;
+    int64_t  ne01;
+    int64_t  ne02;
+    int64_t  ne03;
```
```suggestion
    int32_t  ne00;
    int32_t  ne01;
    int32_t  ne02;
    int32_t  ne03;
```

ggerganov commented on ggml/src/ggml-metal/ggml-metal-impl.h:
```diff
@@ -831,6 +831,25 @@ typedef struct {
+    int64_t  ne0;
+    int64_t  ne1;
+    int64_t  ne2;
+    int64_t  ne3;
```
```suggestion
    int32_t  ne0;
    int32_t  ne1;
    int32_t  ne2;
    int32_t  ne3;
```

gabe-l-hart committed:
fix: Type fixes

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>
Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

gabe-l-hart committed:
feat: Add softplus for metal

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add EXPM1 for metal

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add FILL for metal

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
refactor: Branchless version of tri using _ggml_vec_tri_cmp as a mask

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov commented on ggml/src/ggml-metal/ggml-metal.metal:
```diff
@@ -1943,6 +1989,77 @@ typedef decltype(kernel_cumsum_add<float>) kernel_cumsum_add_t;
+        uint3   tgpig[[threadgroup_position_in_grid]],
+        ushort3 tpitg[[thread_position_in_threadgroup]],
+        ushort  sgitg[[simdgroup_index_in_threadgroup]],
+        ushort  tiisg[[thread_index_in_simdgroup]],
```
These are unused

ggerganov reviewed (approved):
Need to adapt to the changes after #17739

gabe-l-hart committed:
fix: Remove unused arguments

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
Merge remote-tracking branch 'origin/master' into ggml-cumsum-tri

* origin/master:
CUDA: generalized (mma) FA, add Volta support (#17505)
chat : reserve memory in compute_diffs and improve naming (#17729)

gabe-l-hart committed:
Merge remote-tracking branch 'origin/master' into ggml-cumsum-tri

* origin/master:
server: strip content-length header on proxy (#17734)
server: move msg diffs tracking to HTTP thread (#17740)
examples : add missing code block end marker [no ci] (#17756)
common : skip model validation when --help is requested (#17755)
ggml-cpu : remove asserts always evaluating to false (#17728)
convert: use existing local chat_template if mistral-format model has one. (#17749)
cmake : simplify build info detection using standard variables (#17423)
ci : disable ggml-ci-x64-amd-* (#17753)
common: use native MultiByteToWideChar (#17738)
metal : use params per pipeline instance (#17739)
llama : fix sanity checks during quantization (#17721)
build : move _WIN32_WINNT definition to headers (#17736)
build: enable parallel builds in msbuild using MTT (#17708)
ggml-cpu: remove duplicate conditional check 'iid' (#17650)
Add a couple of file types to the text section (#17670)
convert : support latest mistral-common (fix conversion with --mistral-format) (#17712)
Use OpenAI-compatible `/v1/models` endpoint by default (#17689)
webui: Fix zero pasteLongTextToFileLen to disable conversion being overridden (#17445)

gabe-l-hart committed:
refactor: Use select instead of branch for softplus non-vec

Branch: ggml-cumsum-tri

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>
